src/Components/Web/src/WebRenderer.cs

```diff
@@ -34,25 +34,50 @@ public abstract class WebRenderer : Renderer
         JSComponentInterop jsComponentInterop)
         : base(serviceProvider, loggerFactory)
     {
-        _serviceProvider = serviceProvider;
         _interopMethodsReference = DotNetObjectReference.Create(
             new WebRendererInteropMethods(this, jsonOptions, jsComponentInterop));
 
         // Supply a DotNetObjectReference to JS that it can use to call us back for events etc.
         jsComponentInterop.AttachToRenderer(this);
-        var jsRuntime = _serviceProvider.GetRequiredService<IJSRuntime>();
-        jsRuntime.InvokeVoidAsync(
+        var jsRuntime = serviceProvider.GetRequiredService<IJSRuntime>();
+        _attachTask = jsRuntime.InvokeAsync<int>(
```

Comment:
If we want to avoid this dance, we can change the Renderer ID for a GUID instead of an int.

Reply:
I would avoid doing this inside the constructor and instead have a static factory method that we can call to initialize the renderer. If we need to / want to, we can have an interface for it, but I don't think we need it. It avoids the whole task "dance".

Additionally, for Blazor server we can pass in the ID directly on the call to Blazor Start. 

We could similarly do something with web assembly, maybe with an environment variable or similar that we could read from config (Plugging in the environment variables provider). I believe this is possible since we have "harmonized" the boot process.

Reply:
Yeah this is a bit awkward. There's nothing objectively incorrect about starting a task from a constructor, but it definitely pushes up against the boundaries of normal constructor behavior.

The factory method option is certainly a viable possibility, albeit with its own drawbacks (if you're not supposed to use the constructor any more, this becomes a viral concern since the same would be true for any derived classes). Right now we could get away with the factory method as only a minor breaking change.

However @javiercn's points about other approaches are also promising. Do we really need to support registering an arbitrary number of renderers dynamically? What limitations would we face if we just said there would only be at most one of a given type, and we have some convention like 0=server, 1=webassembly? Is it enough just for these two type not to clash?

Reply:
Actually, you bring up a good point. We can use a HILO strategy, where each mode gets a different "HI" value, and we use different LO values. To be clear, if the render mode is an Int32, take the two most significant bytes and give 0 to Server and 1 to Webassembly, (2 to Webview), and use 0 for the two least significant bytes. If in the future we want to support multiple renderers of the same type, each one will just "rev" the ID.

Reply:
> If we want to avoid this dance, we can change the Renderer ID for a GUID instead of an int.

This could work, but wouldn't it mean we're serializing and sending a much larger ID to JavaScript each time we render?

> What limitations would we face if we just said there would only be at most one of a given type, and we have some convention like 0=server, 1=webassembly? Is it enough just for these two type not to clash?

This was the approach I took during prototyping, and it works, but I decided not to do that in this PR because:
* We have to keep those hard-coded IDs in sync between .NET and JS code
* We have to remember to assign new a renderer ID to each execution model we add in the future
* It makes it more difficult to enable having multiple renderers per platform, in case that ever becomes useful (although I recognize Javier's HILO suggestion solves that)

> We can use a HILO strategy, where each mode gets a different "HI" value, and we use different LO values. To be clear, if the render mode is an Int32, take the two most significant bytes and give 0 to Server and 1 to Webassembly, (2 to Webview), and use 0 for the two least significant bytes. If in the future we want to support multiple renderers of the same type, each one will just "rev" the ID.

I'm not sure I'm totally convinced yet that these alternate approaches are better than what we currently have. Even if we were to change the way to allocate renderer IDs, we would still have to call `attachWebRendererInterop` from .NET to provide JS with the dependencies it needs to perform rendering. So, we'd still have the problem of figuring out the right way make that invocation; it would still be problematic to attempt to render before that call completes, even if the .NET side doesn't rely on the return value of `attachWebRendererInterop` to assign a renderer ID. If we have to do the JS interop call anyway, what are we losing by using it to also allocate a renderer ID?

I do totally agree that it's not ideal for the `WebRenderer` constructor to initiate a JS interop call. I think the factory approach would be the best way to handle it, but given that it would be a breaking change (and that we've been invoking the JS interop call in the constructor for a while now already), it seems fine to me to keep things as they are.

Thoughts?

Reply:
> wouldn't it mean we're serializing and sending a much larger ID to JavaScript each time we render?

Yes, would be nice to avoid that.

> and that we've been invoking the JS interop call in the constructor for a while now already

That is a good point - I agree that since this change doesn't actually worsen anything, it's not really necessary to force an architectural or API-breaking change here. Currently this is only an internal detail and if we want to make an API-breaking change to turn this into a factory, there's nothing better about doing that now than any time in the future.

We can see if @javiercn has further practical reasons for changing this but otherwise this looks good to me.